rfcs/0048-BYOC-Marvell-ML-accelerator-integration.md

```diff
@@ -0,0 +1,547 @@
+
+* We are following the existing TVM BYOC framework and example files.
+    * for example: to do IR compositions, to define own IR passes, to mix implementations in Python/C++, and etc.
+
+* We have extended graph_executor_codegen.cc and JSON loader/saver in order to read and write out Marvell specific
+  attributes
+
+* Currently, we haven't spend enough time to under how tvm/rust/cargo requirements and steps. Therefore, we are
+  bypassing the tvm/Jenkinsfile's tests/scripts/task_rust.sh step. We will need help to re-enable the step.
+
+* We like to duplicate the Jenkins environment in order to run tvm/Jenkinsfile as is, but, we ran into many issues.
+  Currently, we have a tvm-like Jenksinsfile environment to only run a subset of test suites using a modified
+  Jenkinsfile.
+
+* We have identified a need to allow a call-back function to be registered when generating Mrvl-BYOC-specific
+  Nodes-JSON file. We are trying to follow TVM Python/CPP-CB style as much as possible. But, since our callback
+  function tvm/src/relay/backend/contrib/mrvl/graph_executor_codegen_mrvl.cc::GetExternalJSON() function is using
+  non-simple argument types, we need help from TVM community to provide suggestions/guidelines in order to make
+  new CB code better to meet TVM community requirements here.
+
+* For one Mrvl-BYOC relay transformation pass, we have identified a need to inject a (global) expr node ID for the
+  RelayExprNode class and its derived classes: Tuple and CallNode, so that during the transformation pass, we can
+  uniquely identify each Tuple or CallNode object. Again, we need help from TVM community to provide
+  suggestions/guidelines here in order to know whether this is one of the best ways to achieve the Mrvl-BYOC need.
+
+* We also identified a need to maintain linkages between (operator-)information described in the original, given
+  pre-trained network model and the code-gen JSON files so that the compiler backend will be able to report user-level
+  (e.g., meaningful-to-user) messages regarding the given pre-trained network. For instance, in the
+  tvm/python/tvm/relay/frontend/onnx.py and common.py files, we can see user-level information being captured using
+  “tvm_custom” related code as in original onnx.py file for the given pre-trained network; but, in common.py, the code
+  later drops the linkage, via attrs.pop(“tvm_custom”), and does not pass the linkage onto the initial relay IR graph.
+  We have a draft solution to maintain linkages between the given pre-trained network model and its relay IR graph
+  (using expr node ID and tvm custom ID, plus, a few utility functions), but would like to know whether the TVM
+  community has any better or work-in-progress resolution.
+
+* When using TVM RPC code to exercise and run inference on a remote-hosted Mrvl ML/AI HW accelerator for the Mrvl
+  subgraph, we ran into one minor issue and have made local TVM RPC enhancement so that, when a TVM RPC client sends
```

Comment:
could you explain the nature of the problem that requires the client to know the absolute path?

Reply:
First, the TVM RPC server choses path for any uploaded file under tmp randomly (which can be good to reduce possible security problem). But, in our use case, we like to have the TVM RPC client to send a “runtime” command to the RPC server side to pre-process the just uploaded file before the file can be consumed autonomously by the RPC server using pre-defined script. We can’t find a way or via a TVM example, which shows how this can be done -- unless the client knows the uploaded file's 
path on server.

Reply:
yes and we will find time to check.

Reply:
BTW, in our use case, we need the server path to be know on the client side so that the client is the master who controls activities to be running on the server side.